Address feedback for the dynamic scheduler post

jmmv opened the pull request:
- Avoid implying that remote execution is problematic: we have picked
  pathological cases for our measurements.
- Clarify that this feature is experimental.
- Warn users about potential correctness issues.
- Clarify that most measurements were taken against Google's internal
  remote execution engine and that they are not indicative of RBE
  performance.

jmmv commented:
@dslomov Could you take a look at this? Also, CC @buchgr .

dslomov commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -48,9 +48,15 @@ There are a couple of prerequisites to use the dynamic spawn scheduler feature:
-We have measured the behavior of the dynamic scheduler on a large iOS app on three  Macs with different local performance characteristics, and also a [large Android app](https://github.com/jin/android-projects#big_connected) on a Linux workstation [using RBE](https://gist.github.com/jin/1fc2543acef7cdbd5618b08579d7210c) for the remote execution backend. The specific details of the build and the m ...
+We have measured the behavior of the dynamic scheduler on a large iOS app on three Macs with different local performance characteristics, and also a [large Android app](https://github.com/jin/android-projects#big_connected) on a Linux workstation. The specific details of the builds and the machines is not relevant.
+
+We have evaluated primarily mobile apps because their critical path contains some actions that are especially costly to run remotely due to the large number of inputs and outputs they have, or the combined sizes of these. *These measurements are not indicative of the average performance you'd get from using any of these remote execution systems as we have purpusely picked some known problematic c ...
```
typo: purpusely

jmmv replied:
Done.

dslomov commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -48,9 +48,15 @@ There are a couple of prerequisites to use the dynamic spawn scheduler feature:
 
 If your setup satisfies the above, all you have to do is specify `--experimental_spawn_scheduler` on your next build. Just make sure to pass this new flag after all others.
 
+*WARNING: As the flag name implies, this feature is experimental. In particular, be aware that this could introduce correctness issues if the local and remote build environments aren't sufficiently similar to yield the same outputs for each action.*
```
maybe link to https://docs.bazel.build/versions/master/backward-compatibility.html#at-a-glance

jmmv replied:
Done.

dslomov commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -1,12 +1,12 @@
 ---
 
-The recently-released [Bazel 0.21](https://blog.bazel.build/2018/12/19/bazel-0.21.html) delivers a new feature that can make your remote builds faster by making use of local resources where possible. Prior to this release, this was already doable to some extent by using manual strategy definitions, but this new feature takes all manual decisions out of the picture.
+The recently-released [Bazel 0.21](https://blog.bazel.build/2018/12/19/bazel-0.21.html) delivers a new feature that can make your builds faster by making use of remote *and* local resources transparently. Prior to this release, this was already doable to some extent by using manual strategy definitions, but this new feature takes all manual decisions out of the picture.
```
I suggest to tone down this paragraph even more to mention that is an experiment, not a "delivered new feature"

jmmv replied:
Done.

buchgr commented:
FYI @ericburnett, @bergsieker

buchgr commented:
Thanks @jmmv! I think someone from RBE should also take a look to make sure they are ok with publishing benchmark results.

EricBurnett commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -71,14 +77,16 @@ iOS | iMac Pro 2018, 18-core | 34s | 414s ↑ | 32s ≈
 iOS | MacBook Pro 2015, 4-core | 39s | 450s ↑ | 36s ≈
 
-As you can see, the remote-only configuration gave us slightly better results than the local-only configuration for clean builds, but the remote-only configuration became unacceptably slow for incremental builds. As expected, though, dynamic execution wins heads-down in all cases: clean builds are many times faster than they are with local or remote execution alone, and incremental builds are ess ...
+As you can see, the remote-only configuration gave us better results than the local-only configuration for clean builds, but the remote-only configuration became unacceptably slow for incremental builds. As expected, though, dynamic execution won heads-down in all cases: clean builds were many times faster than they are with local or remote execution alone, and incremental builds were essentially ...
```
Neat! These are pretty sweet results, nice work :).

EricBurnett replied:
Oh, one more point: if you wait till EoW and then re-test, the remote times should get better across the board: iirc one of the issues with these builds was that they have lots of inputs per action, and our current release has performance improvements landing targeting that use-case.

I wouldn't block publishing before that though.

jmmv replied:
Cool. @jin can retest Android when he's back, which should be EoW as well :) iOS is purely internal so your fixes wouldn't fix that :-/

EricBurnett replied:
Yeah, sorry, I meant to say "RBE times" vs all remote. And to be clear, I don't know what the magnitude of the impact will be - certainly not enough to make RBE competitive with local for incremental android builds I don't think, but hopefully not as painfully far :).

EricBurnett commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -71,14 +77,16 @@ iOS | iMac Pro 2018, 18-core | 34s | 414s ↑ | 32s ≈
+As you can see, the remote-only configuration gave us better results than the local-only configuration for clean builds, but the remote-only configuration became unacceptably slow for incremental builds. As expected, though, dynamic execution won heads-down in all cases: clean builds were many times faster than they are with local or remote execution alone, and incremental builds were essentially ...
 
-We expect the speedup to be more prominent as builds get larger, as seen in the difference between the iOS and Android builds.
+We expect the speedup to be more prominent as builds get larger and bottlenecked on single actions, as seen in the difference between the iOS and Android builds.
```
Which speedup? This sentence is a little ambiguous to me.

I generally expect Remote to get better as builds get larger, since larger builds are generally less critical-path bound and more throughput-bound. But that doesn't sound like what you're saying?

jmmv replied:
The experience with iOS doesn't match your expectation. I think the problem is that larger builds balloon in the number of inputs per action so there end up being more actions that are much more costly (and linking/bundling become even slower). This is a pathological case with iOS support in this particular configuration thoug. Also I think what I had in mind was the impact on incremental builds where we only have the critical path.

But... better remove this altogether. I'm not sure what I'd claim :P

jmmv commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -48,9 +48,15 @@ There are a couple of prerequisites to use the dynamic spawn scheduler feature:
 
 If your setup satisfies the above, all you have to do is specify `--experimental_spawn_scheduler` on your next build. Just make sure to pass this new flag after all others.
 
+*WARNING: As the flag name implies, [this feature is experimental](https://docs.bazel.build/versions/master/backward-compatibility.html#at-a-glance). In particular, be aware that this could introduce correctness issues if the local and remote build environments aren't sufficiently similar to yield the same outputs for each action.*
```
Well, for iOS we are using this feature inside Google by default... but we have the benefit of tightly controlling the local execution environment. But iOS in the public may be a different story.

There are many caveats here. If people use checked-in tools, for example, they should be able to mitigate most differences right?

I have reworded this to soften the tone and suggest using a container.

EricBurnett replied:
LG, thanks.

> There are many caveats here. If people use checked-in tools, for example, they should be able to mitigate most differences right?

Yes, but unfortunately that is almost never done outside of Google (excepting a few language rules like Go), and in many cases it _can't_ be done (Git doesn't natively handle inline binaries well, though various folk are trying to improve that). So I wouldn't want the blog post to rely on that. But I think it's clear now without assuming that, so no concerns from me.

jmmv replied:
OK cool. Resolving thread.

EricBurnett commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -23,17 +23,17 @@ Action mnemonics are statically mapped to strategies via command-line flags. Whe
 You'd think that this is not a big deal though---if remote execution offers sufficient parallelism, the fact that each action has higher latency is compensated by the fact that we have a much higher throughput than what you'd ever get with pure local execution. And you'd think right---this has been the case for a long time, at least within Google, because the overheads of remote execution were no ...
 
-But this model breaks down when the cost of remote execution is non-trivial. For example, imagine if a link action that takes about 30 seconds on the local machine requires up to 10 minutes on a remote machine (and, yes, we have painfully observed this due to file transfer artifacts). And notice that I said *link* action---linking happens in every single build, and linking is almost-certainly in  ...
+But this model breaks down when the cost of remote execution is non-trivial. For example, imagine if a link action that takes about 30 seconds on the local machine requires up to 10 minutes on a remote machine (and, yes, we have painfully observed this within Google due to the way we handle specific cases of this action). And notice that I said *link* action---linking happens in every single buil ...
```
Thanks! LGTM, I like how it came out.

bazelpm commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -23,17 +23,23 @@ Action mnemonics are statically mapped to strategies via command-line flags. Whe
 ## The problem with remote execution
 
-Unfortunately, in the case of remote execution, sending *all* actions through the `remote` strategy is inefficient. Each remote action comes with a cost: there is the overhead of the RPCs to contact with the remote service and there is also the more-significant overhead of uploading inputs and downloading outputs (even if we can [maybe optimize the latter](https://docs.google.com/document/d/11m5A ...
+In the case of a fully remote build, sending *all* actions through the `remote` strategy can be inefficient. Each remote action comes with a cost: there is the overhead of the RPCs to contact with the remote service and there is also the more-significant overhead of uploading inputs and downloading outputs (even if we can [maybe optimize the latter](https://docs.google.com/document/d/11m5AkWjigMg ...
```
Two small writing nits: It should be "more significant" instead of "more-significant". I would also say "even if we may be able to optimize the latter" rather than "can maybe".

jmmv replied:
Done.

bazelpm commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -23,17 +23,23 @@ Action mnemonics are statically mapped to strategies via command-line flags. Whe
+
+1.  When the compiler is so fast that it's cheaper to run the action locally than remotely, and said actions are in the critical path. Examples include the Go compiler or the Java persistent workers.
+
+The introduced delays in incremental builds kill the edit/build/test cycle, so some users affected by this problem have rightfully objected to using remote execution.
```
I would say "understandably" rather than "rightfully".

jmmv replied:
Done.

bazelpm commented on _posts/2019-02-01-dynamic-spawn-scheduler.md:
```diff
@@ -71,14 +83,14 @@ iOS | iMac Pro 2018, 18-core | 34s | 414s ↑ | 32s ≈
 ## Future work and credits
 
-There still are some problems to be resolved, of course. The major one is that local workers do not support cancellations ([#614](https://github.com/bazelbuild/bazel/issues/614])), and this can result in a high build timing variance depending on how the actions are scheduled. Other minor issues include: an artificial delay before running actions locally ([#7327](https://github.com/bazelbuild/baze ...
+There still are some problems to be resolved, of course, and is why this feature is still **experimental**. The major one is that *using this feature can lead to correctness issues unless you can guarantee that the local build environment matches the remote one* (if you cannot do this, avoid using this feature). Minor issues include the fact that local workers do not support cancellations ([#614] ...
```
It's unclear from the header that this section is really about caveats. What do you think about labeling it "Caveats" or "Considerations before using"? It also seems like credits could be a different section.

jmmv replied:
Renamed and added credits. I don't really like such a section at the end but can't think of a better title.

jmmv commented:
I think I have addressed all feedback,and in particular from @EricBurnett (RBE). I think this is ready to merge with an approval from any of you -- and the sooner we amend the post, the better.

dslomov commented:
I am giving people until 23:00 CEST = 15:00 NYC time to comment.

EricBurnett commented:
LGTM from me, though I'll leave it to bazel folk to formally Approve and merge this.

jmmv commented:
@dslomov It's already past the deadline you gave and nobody has raised objections. In fact, they have LGTMd the post. So I consider this good to go. I'll dismiss your request to make changes in about an hour, just in case you are a little bit late.

bazelpm commented:
@jmmv The post LGTM as well—I mentioned that over chat but should mention it here, too.

dslomov commented:
hey it's only 22:08 CEST here

On Wed, 6 Feb 2019, 21:47 David Jacobs, <notifications@github.com> wrote:

> @jmmv <https://github.com/jmmv> The post LGTM as well—I mentioned that
> over chat but should mention it here, too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel-blog/pull/152#issuecomment-461181315>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AExrX6iLxVeJC1Bxk6eQkazRWWIxBdSQks5vKz9ugaJpZM4ahZMq>
> .
>

dslomov commented:
(I guess I messed up timezone conversions)

On Wed, 6 Feb 2019, 22:08 Dmitry Lomov, <dslomov@google.com> wrote:

> hey it's only 22:08 CEST here
>
> On Wed, 6 Feb 2019, 21:47 David Jacobs, <notifications@github.com> wrote:
>
>> @jmmv <https://github.com/jmmv> The post LGTM as well—I mentioned that
>> over chat but should mention it here, too.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/bazelbuild/bazel-blog/pull/152#issuecomment-461181315>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AExrX6iLxVeJC1Bxk6eQkazRWWIxBdSQks5vKz9ugaJpZM4ahZMq>
>> .
>>
>

jmmv commented:
Heh, anyway... is there a point in waiting an extra hour?

dslomov commented:
no it's fine, go ahead and submit ;)

On Wed, 6 Feb 2019, 22:10 Julio Merino, <notifications@github.com> wrote:

> Heh, anyway... is there a point in waiting an extra hour?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel-blog/pull/152#issuecomment-461188538>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AExrX32vNjEJOVdBGrrFBEHLeD1UhCbvks5vK0TAgaJpZM4ahZMq>
> .
>

jmmv commented:
Thanks to everyone for the feedback! And let's put a process in place to coordinate better on any future posts.
